Add Daffodil schema template

stevedlawrence opened the pull request:
- Create a Giter8 template repository for easily creating Daffodil
  schema projects. Everything in src/main/g8/ uses template syntax. Note
  that all the files in this directory are Apache Licensed, but the
  licenses are in g8 comments so that when files are generated they are
  not licensed. This lets the user determine what license they want
  their schema to use.
- Also includes .asf.yaml and .github files to configure GithHub
  settings and continuous integration

DAFFODIL-2144

tuxji commented on src/main/g8/.gitattributes:
```diff
@@ -0,0 +1,18 @@
+# limitations under the License.
+$endif$
+
+/src/test/resources/**/data/** text=false
```
You might want to remove the blank line at 17 or move it before line 16 so that the generated .gitattributes doesn't start with a blank line.  Also, should the directory at 18 start with a slash - shouldn't it be a relative path (`src/test/resources/**/data/**`)?

stevedlawrence replied:
Yep, I'll fix all the blank line issues.

I think the beginning slash actually doesn't make a difference in this pattern, but it does look at bit weird. I'll remove it.

tuxji commented on src/main/g8/.gitignore:
```diff
@@ -0,0 +1,18 @@
+# limitations under the License.
+!$
+
+target
```
You might want to remove the blank line at 17 or move it before line 16 so the generated .gitignore doesn't start with a blank line.

tuxji commented on src/main/g8/build.sbt:
```diff
@@ -0,0 +1,34 @@
+// limitations under the License.
+!$
+
+name := "dfdl-$name;format="norm"$"
```
Ditto about the blank line at 17.

tuxji commented on src/main/g8/project/build.properties:
```diff
@@ -0,0 +1,18 @@
+# limitations under the License.
+!$
+
+sbt.version=1.3.9
```
Ditto about line 17

tuxji commented on src/main/g8/src/test/resources/$package$/$name__camel$/infosets/test_01.$name__camel$.xml:
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
You might want to remove the unnecessary whitespace character before the `?>`.

stevedlawrence replied:
Good eye, will fix.

tuxji commented on src/main/g8/src/test/resources/$package$/$name__camel$/infosets/test_01.$name__camel$.xml:
```diff
@@ -0,0 +1,21 @@
+-->
+!$
+
+<$name;format="camel"$:$name;format="Camel"$ xmlns:$name;format="camel"$="urn:$name;format="camel"$" />
```
Ditto about blank line at 20.

tuxji commented on src/main/g8/src/test/scala/$package$/$name__camel$/Test$name__Camel$.scala:
```diff
@@ -0,0 +1,38 @@
+
+class Test$name;format="Camel"$ {
+
+  import Test$name;format="Camel"$._
```
Maybe remove blank line at 33?

stevedlawrence replied:
Looks like the convention we use in Daffodil test scala files is to have the blank line before the import, I think I might keep this one just to be consistent with that.

tuxji commented on src/main/g8/src/test/scala/$package$/$name__camel$/Test$name__Camel$.scala:
```diff
@@ -0,0 +1,38 @@
+
+  import Test$name;format="Camel"$._
+
+  @Test def test_$name;format="camel"$_01() = { runner.runOneTest("test_$name;format="camel"$_01") }
```
You may want to give the method a type (`: Unit = {`).

stevedlawrence replied:
Yep, good catch

tuxji commented on src/main/g8/src/test/scala/$package$/$name__camel$/Test$name__Camel$.scala:
```diff
@@ -0,0 +1,38 @@
+
+  @Test def test_$name;format="camel"$_01() = { runner.runOneTest("test_$name;format="camel"$_01") }
+  
+}
```
Maybe remove blank line at 37?

tuxji commented on src/test/g8/test:
```diff
@@ -0,0 +1,21 @@
+$ exists target/dfdl-format-0.1.0-SNAPSHOT.jar
+
+# make sure the tests pass
+> test
```
If this file is supposed to be a template, I didn't see any file like it in the generated project.  Is the sbt test command in `.github/workflows/main.yml` supposed to call this file?

stevedlawrence replied:
This is a [scripted](https://www.scala-sbt.org/1.x/docs/Testing-sbt-plugins.html) test file. These are usually used for testing plugins, but it can also be used for testing g8. Essentially what happens is g8 will run "sbt new" on itself, then use this file to run sbt commands and commands that test that those sbt commands work. So this essentially does "sbt new "..., then "sbt package" and makes sure it builds the jar file, then "sbt test" and makes sure those tests pass.

mbeckerle replied:
Please add comment akin to what you just said, into the file, or if you prefer, into the README or somewhere.

stevedlawrence replied:
Added a comment in the README, I figure more people are going to see that there.

Also, I learned that this only allows  a single sanity check test. Kindof dissappointing, but it's a limitation of the giter8 plugin we use.

tuxji commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,53 @@
+        run: sbt ++${{ matrix.scala_version }} ratCheck
+
+      - name: Test
+        run: sbt ++${{ matrix.scala_version }} test
```
Is this how the `src/test/g8/test` file gets used?  Might want to change name to `name: Run Tests`.

Are you enabling GitHub Actions?  This pull request didn't run any CI job after you created it.

stevedlawrence replied:
Yeah, this "sbt test" will run the scripted g8 test mentioned before. Ill rename the action to make it more clear.

I think maybe the GitHub action won't be enabled until it's in the master branch, maybe? You can look see what it will look like on my fork: https://github.com/stevedlawrence/daffodil-schema.g8/runs/1982074448

tuxji commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,53 @@
+      matrix:
+        java_version: [ 8, 11, 15 ]
+        scala_version: [ 2.12.11 ]
+        os: [ 'ubuntu-latest' ]
```
You can say just `[ ubuntu-latest ]` - the quotes aren't mandatory.

tuxji commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,53 @@
+      ############################################################
+
+      - name: Checkout Repository
+        uses: actions/checkout@v2.0.0
```
Why `@v2.0.0` instead of `@v2`?  This prevents the workflow from using any later version in the v2 series.

stevedlawrence replied:
I know we've ran into an issue at one point where github made an update to the action that had the same major version and it actually broke something. To avoid that, we changed our actions to specify a specific version. I'm fine updating to a newer version if one exist, but to prevent random breakages, I sort of like to stick with known good versions.

mbeckerle replied:
Open ended version dependencies have been used for supply-chain security attacks recently. The only secure practice is to specify exact version numbers of all dependencies. The hack is called "Dependency Hijacking"

tuxji commented on NOTICE:
```diff
@@ -0,0 +1,5 @@
+Copyright 2021 The Apache Software Foundation
+
+This product includes software developed at
+The Apache Software Foundation (http://www.apache.org/).
```
This NOTICE doesn't mention any specific software - consider removing this file or mentioning the giter8 software.

stevedlawrence replied:
The "software" in this case is all this source code. I believe this is the standard boilerplate that goes in Apache notice files.

stevedlawrence replied:
FYI, here's there reference: https://www.apache.org/legal/src-headers.html#notice

tuxji reviewed (changes requested):
Some changes suggested, looks fine otherwise.

tuxji reviewed (approved):
+1

Thanks for answering my questions!  I'm sure the push event after the request is merged will trigger a CI job although I was surprised the initial pull request event hadn't triggered any job.

mbeckerle commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,53 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+name: Daffodil Schema Template CI
```
Suggest Daffodil DFDL Schema Template CI

stevedlawrence replied:
Should be updated everywhere now. Is it worth bothering INFRA to rename the repo to "daffodil-dfdl-schema.g8" or do you think "daffodil-schema.g8" is fine? I do like the simplicity of "daffodil-schema.g8"

mbeckerle replied:
Agreed. I think  we should not rename the repo, just the more verbose places in the template repo where we can be more specific.

mbeckerle commented on NOTICE:
```diff
@@ -0,0 +1,5 @@
+Apache Daffodil Schema Template
```
Suggest Apache Daffodil DFDL Schema Template

mbeckerle commented on README.md:
```diff
@@ -0,0 +1,65 @@
+limitations under the License.
+-->
+
+# Apache Daffodil Schema Template
```
And here too.

mbeckerle commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+## License
+
+Apache Daffodil Schema Template is licensed under the [Apache License, v2.0].
```
and here

mbeckerle commented on README.md:
```diff
@@ -0,0 +1,65 @@
+
+### Properties
+
+The following properties are used to configured the generated schema project:
```
DFDL schema

mbeckerle commented on project/plugins.sbt:
```diff
@@ -0,0 +1,18 @@
+
+addSbtPlugin("org.foundweekends.giter8" %% "sbt-giter8" % "0.13.1")
+
+addSbtPlugin("org.musigma" % "sbt-rat" % "0.7.0")
```
If a user creates a DFDL schema project, does their schema now get ratchecks run on it? 
The user's schema isn't necessarily ASL licensed at all, so ratcheck seems inappropriate. I'm not opposed to having the plugin here, and encouraging use of the ASL license for these projects, but for example I don't think CI should be running ratcheck on user schema projects unless we are able to somehow detect the presence of the ASL license as having been selected (or at least not unselected) by the author of the schema project.

stevedlawrence replied:
Only files in src/main/g8 are end up in the resulting stuff. This file defines plugins for just this repo. Since there isn't a plugins file in src/main/g8, there will be no plugins in the generated sbt project.

mbeckerle commented on src/main/g8/src/test/resources/$package$/$name__camel$/Test$name__Camel$.tdml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="UTF-8"?>
+$!
```
why do so many of these files have "_Camel" in their template names?

stevedlawrence replied:
It's g8 syntax to format the name property. name__Camel means take the "name" property and make it camel case with the first letter capitalized. name__camel means take the "name" property and make it camel case with the first letter lowercase.

So if you generate a project with name "My Great Format", in creates myGreatFormat/TestMyGreatFormat.tdml

There's lots of other formatting capabilities: http://www.foundweekends.org/giter8/formatting.html

Note that the __ stuff is used only for directory/file names, since windows doesn't like the ``;format="..."`` syntax that used inside files.

mbeckerle replied:
Ah,  I thought it was some holdover from using this tool for Apache Camel or something. This makes perfect sense. Pretty clever actually.

mbeckerle reviewed (approved):
+1 I suggest we change all uses of "Daffodil Schema" to "Daffodil DFDL Schema", a few other minor questions, that should probably be answered in the README or other how-it-works documentation.  I'm unclear on why ratcheck is part of this.

stevedlawrence committed:
Add Daffodil schema template

- Create a Giter8 template repository for easily creating Daffodil DFDL
  schema projects. Everything in src/main/g8/ is the actual template.
  Note that all the files in this directory are Apache Licensed, but the
  licenses are in g8 comments so that when files are generated they are
  not licensed. This lets the user determine what license they want
  their schema to use.
- Also includes .asf.yaml and .github files to configure GithHub
  settings and continuous integration

DAFFODIL-2144
